src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchTree.cs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Configuration
+{
+    internal class SearchTree
+    {
+        private readonly SearchNode _root;
+
+        internal SearchTree(PackageNamespacesConfiguration configuration)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            _root = new SearchNode();
+
+            foreach (var namespacePerSource in configuration.Namespaces)
+            {
+                foreach (var namespaceId in namespacePerSource.Value)
+                {
+                    Add(namespacePerSource.Key, namespaceId);
+                }
+            }
+        }
+
+        private void Add(string packageSourceKey, string namespaceId)
+        {
+            SearchNode currentNode = _root;
+
+            if (namespaceId == null)
+            {
+                throw new ArgumentNullException(nameof(namespaceId));
+            }
+
+            // To prevent from unwanted behaviour.
+            if (namespaceId.Length > PackageNamespacesConfiguration.PackageIdMaxLength)
+            {
+                throw new ArgumentOutOfRangeException(nameof(namespaceId));
+            }
+
+            packageSourceKey = packageSourceKey.ToLowerInvariant().Trim();
+            namespaceId = namespaceId.ToLowerInvariant().Trim();
+
+            for (int i = 0; i < namespaceId.Length; i++)
+            {
+                char c = namespaceId[i];
+
+                if (c == '*')
+                {
+                    // break here since seeing * means end of expression.
+                    currentNode.IsGlobbing = true;
+                    break;
+                }
+
+                if (!currentNode.Children.ContainsKey(c))
+                {
+                    currentNode.Children[c] = new SearchNode();
+                }
+
+                currentNode = currentNode.Children[c];
+            }
+
+            if (string.IsNullOrEmpty(currentNode.NamespaceId))
+            {
+                currentNode.NamespaceId = namespaceId;
+            }
+
+            if (currentNode.PackageSources == null)
+            {
+                currentNode.PackageSources = new List<string>();
+            }
+
+            currentNode.PackageSources.Add(packageSourceKey.ToLowerInvariant());
```

Comment:
You already made a lower-case string [above](https://github.com/NuGet/NuGet.Client/pull/4039/files#diff-c4e242bf264758dcb10fd006bd06de268804d46fd2dc83fb8b1843b36dae9c72R46).  Try to reuse it instead of creating a new string.

Besides, why not just keep the package source key as-is?  By lower-casing it here, later we'll log the lower-case package source name.  I don't see why we wouldn't preserve the casing from NuGet.config where we can, [especially when displaying it back to a user](https://github.com/NuGet/NuGet.Client/pull/4039/files#diff-34f952aa1e186f8b6d14281e0c35727023138e5aa7150dc30655ea2176aab728R108-R109).

Reply:
I agree. Lower casing package source key doesn't make much sense.